File: backend/ai-service/app/schemas.py
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, field_validator, model_validator


class WorkflowName(str, Enum):
    assistant_chat = "assistant_chat"
    goal_plan_generation = "goal_plan_generation"
    vibe_feedback = "vibe_feedback"
    syllabus_import = "syllabus_import"


class AIRunRequest(BaseModel):
    workflow: WorkflowName
    payload: Dict[str, Any] = Field(default_factory=dict)


class AIError(BaseModel):
    code: str
    message: str


class AIErrorResponse(BaseModel):
    error: AIError


class AIRunResponse(BaseModel):
    workflow: WorkflowName
    result: Dict[str, Any]
    draftID: Optional[str] = None


class AssistantChatPayload(BaseModel):
    message: str = Field(min_length=1, max_length=4000)
    timezone: str = Field(min_length=1, max_length=80)
    currentScreen: Optional[str] = Field(default=None, max_length=120)
    date: Optional[str] = None
    contextHints: Dict[str, Any] = Field(default_factory=dict)


class DraftAction(BaseModel):
    type: str = Field(min_length=1, max_length=80)
    title: str = Field(min_length=1, max_length=160)
    dueAt: Optional[str] = None
    reason: str = Field(min_length=1, max_length=400)


class AssistantChatResult(BaseModel):
    message: str = Field(min_length=1, max_length=1600)
    draftActions: List[DraftAction] = Field(default_factory=list, max_length=3)


class GoalDetails(BaseModel):
    title: str = Field(min_length=1, max_length=160)
    description: str = Field(default="", max_length=4000)


class GoalPlanPayload(BaseModel):
    goalID: Optional[str] = Field(default=None, min_length=1, max_length=200)
    goal: Optional[GoalDetails] = None
    timelineWeeks: int = Field(ge=1, le=52)
    startDate: str = Field(min_length=1, max_length=80)
    timezone: str = Field(min_length=1, max_length=80)

    @model_validator(mode="after")
    def require_goal_reference(self) -> "GoalPlanPayload":
        if not self.goalID and not self.goal:
            raise ValueError("Either goalID or goal details are required.")
        return self


class GoalMilestone(BaseModel):
    title: str = Field(min_length=1, max_length=160)
    dueDate: str = Field(min_length=1, max_length=80)
    description: str = Field(min_length=1, max_length=600)


class GoalNextAction(BaseModel):
    title: str = Field(min_length=1, max_length=160)
    estimatedMinutes: int = Field(ge=1, le=480)
    priority: str

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, value: str) -> str:
        if value not in {"low", "medium", "high"}:
            raise ValueError("priority must be low, medium, or high")
        return value


class GoalPlanResult(BaseModel):
    summary: str = Field(min_length=1, max_length=1000)
    milestones: List[GoalMilestone] = Field(default_factory=list)
    nextActions: List[GoalNextAction] = Field(min_length=3, max_length=5)


class VibeFeedbackPayload(BaseModel):
    reflectionText: str = Field(min_length=1, max_length=4000)
    timezone: str = Field(min_length=1, max_length=80)
    recentContext: Optional[Dict[str, Any]] = None


class VibeFeedbackResult(BaseModel):
    feedback: str = Field(min_length=1, max_length=1000)
    needs_escalation: bool = False


class SyllabusImportPayload(BaseModel):
    extractedText: str = Field(min_length=1, max_length=120000)
    currentDate: Optional[str] = None
    timezone: str = Field(min_length=1, max_length=80)


class SyllabusAssignment(BaseModel):
    title: str = Field(min_length=1, max_length=240)
    type: Optional[str] = Field(default=None, max_length=120)
    dueDate: Optional[str] = None
    confidence: str
    sourceText: str = Field(min_length=1, max_length=1000)

    @field_validator("confidence")
    @classmethod
    def validate_confidence(cls, value: str) -> str:
        if value not in {"low", "medium", "high"}:
            raise ValueError("confidence must be low, medium, or high")
        return value


class SyllabusCourse(BaseModel):
    name: str = Field(min_length=1, max_length=240)
    instructor: Optional[str] = Field(default=None, max_length=160)
    assignments: List[SyllabusAssignment] = Field(default_factory=list)


class SyllabusWarning(BaseModel):
    message: str = Field(min_length=1, max_length=600)
    sourceText: Optional[str] = Field(default=None, max_length=1000)


class SyllabusImportResult(BaseModel):
    courses: List[SyllabusCourse] = Field(default_factory=list)
    warnings: List[SyllabusWarning] = Field(default_factory=list)

File: backend/functions/src/ai/schemas.ts
import { z } from "genkit";

export const aiWorkflowSchema = z.enum([
  "assistant_chat",
  "goal_plan_generation",
  "vibe_feedback",
  "syllabus_import"
]);

const looseObjectSchema = z.record(z.any()).default({});

export const assistantChatPayloadSchema = z.object({
  message: z.string().trim().min(1).max(2000),
  timezone: z.string().trim().min(1).max(64),
  currentScreen: z.string().trim().max(128).optional().nullable(),
  date: z.string().trim().max(64).optional().nullable(),
  contextHints: looseObjectSchema
});

export const assistantDraftActionSchema = z.object({
  type: z.enum(["goal_plan", "planner_adjustment"]),
  title: z.string().trim().min(1),
  dueAt: z.string().trim().nullable(),
  reason: z.string().trim().min(1)
});

export const assistantChatResultSchema = z.object({
  message: z.string().trim().min(1),
  draftActions: z.array(assistantDraftActionSchema),
  degraded: z.boolean().optional()
});

export const goalDetailsSchema = z.object({
  title: z.string().trim().min(1).max(160),
  description: z.string().trim().max(2000).default("")
});

export const goalPlanGenerationPayloadSchema = z.object({
  goalID: z.string().trim().min(1).max(128).optional().nullable(),
  goal: goalDetailsSchema.optional().nullable(),
  timelineWeeks: z.number().int().positive().max(104),
  startDate: z.string().trim().min(1).max(64),
  timezone: z.string().trim().min(1).max(64)
});

export const goalMilestoneSchema = z.object({
  title: z.string().trim().min(1),
  dueDate: z.string().trim().min(1),
  description: z.string().trim().min(1)
});

export const goalNextActionSchema = z.object({
  title: z.string().trim().min(1),
  estimatedMinutes: z.number().int().positive(),
  priority: z.enum(["low", "medium", "high"])
});

export const goalPlanGenerationResultSchema = z.object({
  summary: z.string().trim().min(1),
  milestones: z.array(goalMilestoneSchema),
  nextActions: z.array(goalNextActionSchema).min(3).max(5),
  degraded: z.boolean().optional()
});

export const vibeFeedbackPayloadSchema = z.object({
  reflectionText: z.string().trim().min(1).max(4000),
  timezone: z.string().trim().min(1).max(64),
  recentContext: looseObjectSchema.optional()
});

export const vibeFeedbackResultSchema = z.object({
  feedback: z.string().trim().min(1),
  needs_escalation: z.boolean(),
  degraded: z.boolean().optional()
});

export const syllabusImportPayloadSchema = z.object({
  extractedText: z.string().trim().min(1).max(120000),
  currentDate: z.string().trim().max(64).optional().nullable(),
  timezone: z.string().trim().min(1).max(64),
  sourceName: z.string().trim().max(256).optional().nullable(),
  uploadedFilePath: z.string().trim().max(1024).optional().nullable()
});

export const syllabusAssignmentSchema = z.object({
  title: z.string().trim().min(1),
  type: z.string().trim().nullable(),
  dueDate: z.string().trim().nullable(),
  confidence: z.enum(["low", "medium", "high"]),
  sourceText: z.string().trim().min(1)
});

export const syllabusCourseSchema = z.object({
  name: z.string().trim().min(1),
  instructor: z.string().trim().nullable(),
  assignments: z.array(syllabusAssignmentSchema)
});

export const syllabusWarningSchema = z.object({
  message: z.string().trim().min(1),
  sourceText: z.string().trim().nullable()
});

export const syllabusImportResultSchema = z.object({
  courses: z.array(syllabusCourseSchema),
  warnings: z.array(syllabusWarningSchema),
  degraded: z.boolean().optional()
});

export const aiRunRequestSchema = z.discriminatedUnion("workflow", [
  z.object({
    workflow: z.literal("assistant_chat"),
    payload: assistantChatPayloadSchema
  }),
  z.object({
    workflow: z.literal("goal_plan_generation"),
    payload: goalPlanGenerationPayloadSchema
  }),
  z.object({
    workflow: z.literal("vibe_feedback"),
    payload: vibeFeedbackPayloadSchema
  }),
  z.object({
    workflow: z.literal("syllabus_import"),
    payload: syllabusImportPayloadSchema
  })
]);

export type AIWorkflow = z.infer<typeof aiWorkflowSchema>;
export type AIRunRequest = z.infer<typeof aiRunRequestSchema>;
export type AssistantChatPayload = z.infer<typeof assistantChatPayloadSchema>;
export type AssistantChatResult = z.infer<typeof assistantChatResultSchema>;
export type GoalPlanGenerationPayload = z.infer<typeof goalPlanGenerationPayloadSchema>;
export type GoalPlanGenerationResult = z.infer<typeof goalPlanGenerationResultSchema>;
export type VibeFeedbackPayload = z.infer<typeof vibeFeedbackPayloadSchema>;
export type VibeFeedbackResult = z.infer<typeof vibeFeedbackResultSchema>;
export type SyllabusImportPayload = z.infer<typeof syllabusImportPayloadSchema>;
export type SyllabusImportResult = z.infer<typeof syllabusImportResultSchema>;
